### PR TITLE
Port scala.quoted.runtime.impl.QuoteMatcher

### PR DESCRIPTION
**Pull request by nicolasstucki**
This will make it simpler to maintain and also improves a bit the performace.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
test performance with #quotes please

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 1]
```diff
@@ -2935,24 +2935,19 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
...
-    val matcher = new Matcher.QuoteMatcher[qctx1.type](qctx1) {
-      def patternHoleSymbol: qctx1.reflect.Symbol = dotc.core.Symbols.defn.QuotedRuntimePatterns_patternHole.asInstanceOf
-      def higherOrderHoleSymbol: qctx1.reflect.Symbol = dotc.core.Symbols.defn.QuotedRuntimePatterns_higherOrderHole.asInstanceOf
-    }
-
     val matchings =
-      if pat1.isType then matcher.termMatch(scrutinee.asInstanceOf[matcher.qctx.reflect.Term], pat1.asInstanceOf[matcher.qctx.reflect.Term])
-      else matcher.termMatch(scrutinee.asInstanceOf[matcher.qctx.reflect.Term], pat1.asInstanceOf[matcher.qctx.reflect.Term])
+      if pat1.isType then QuoteMatcher.termMatch(scrutinee, pat1)(using ctx1)
+      else QuoteMatcher.termMatch(scrutinee, pat1)(using ctx1)
```
Comment (liufengyun) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2940]: This `if/else` seems not useful -- both branches have the same code.

**Review by liufengyun — APPROVED** [review]
LGTM

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: Removed unnecessary duplication

**Review by liufengyun — APPROVED** [review]

**Pull request merged**